homeassistant/components/lunatone/config_flow.py

```diff
@@ -0,0 +1,203 @@
+        step_id = "dali"
+
+        if self._dali_device_scan_task:
+            next_step_id = None
+            if self._dali_device_scan_task.cancelled():
+                _LOGGER.debug("DALI device scan timeout")
+                next_step_id = "failed"
+            elif self._dali_device_scan_task.done():
+                next_step_id = "finish"
+            if next_step_id:
+                self._dali_device_scan_task = None
+                return self.async_show_progress_done(next_step_id=next_step_id)
+
+        if user_input is not None:
+            method = user_input[CONF_SCAN_METHOD]
+            self._dali_device_scan_task = self.hass.async_create_task(
+                self._async_start_dali_device_scan(method)
+            )
+            return self.async_show_progress(
+                step_id=step_id,
+                progress_action="device_scan",
+                progress_task=self._dali_device_scan_task,
+            )
+        return self.async_show_form(
+            step_id=step_id,
+            data_schema=DALI_SCAN_SCHEMA,
+            last_step=False,
+        )
+
+    async def async_step_failed(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle failed config flow."""
+        return self.async_abort(reason="dali_device_scan_timeout")
+
+    async def async_step_finish(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle finishing the config flow."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title=self._title,
+                data={CONF_URL: self._url},
+            )
+        return self.async_show_form()
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration flow initialized by the user."""
+        return await self.async_step_user(user_input)
+
+    async def _async_start_dali_device_scan(self, method: DALIDeviceScanMethod) -> None:
+        auth = Auth(
+            session=async_get_clientsession(self.hass),
+            base_url=self._url,
+        )
+        scan = DALIScan(auth)
+        await scan.async_cancel()
+        if not await self._async_is_dali_device_scan_done(scan):
+            raise asyncio.CancelledError
+
+        if method == DALIDeviceScanMethod.SYSTEM_EXTENSION:
+            start_scan_data = StartScanData()
+        elif method == DALIDeviceScanMethod.NEW_INSTALLATION:
+            start_scan_data = StartScanData(newInstallation=True)
+        else:
+            start_scan_data = StartScanData(noAddressing=True)
+
+        await scan.async_start(start_scan_data)
+        if not await self._async_is_dali_device_scan_done(scan):
+            raise asyncio.CancelledError
+
+    async def _async_is_dali_device_scan_done(self, scan: DALIScan) -> bool:
+        await asyncio.sleep(1)
+        for _ in range(1800):
```

Comment:
lets move this number to a const at the top of this file or a variable in this method, so that we don't have a magic number. Also, isn't 30 min too high?